File: AlgorithmsPackJavaSolutions/Path With Minimum Effort.java
class Solution {
    private int[] dr = {-1, 0, 1, 0};
    private int[] dc = {0, 1, 0, -1};

    private boolean isValid(int r, int c, int[][] heights) {
        if (r < 0 || r >= heights.length)
            return false;
        if (c < 0 || c >= heights[0].length)
            return false;
        return true;
    }

    private boolean dfs(int r, int c, int[][] heights, boolean[][] visited, int diff) {
        if (r == heights.length - 1 && c == heights[0].length - 1)
            return true;

        visited[r][c] = true;

        for (int d = 0; d < 4; d++) {
            int nr = dr[d] + r;
            int nc = dc[d] + c;
            if (!isValid(nr, nc, heights) || visited[nr][nc] || Math.abs(heights[nr][nc] - heights[r][c]) > diff)
                continue;

            if (dfs(nr, nc, heights, visited, diff))
                return true;
        }
        return false;
    }

    private boolean possible(int[][] heights, int diff) {
        int rows = heights.length;
        int cols = heights[0].length;
        boolean[][] visited = new boolean[rows][cols];

        return dfs(0, 0, heights, visited, diff);
    }

    public int minimumEffortPath(int[][] heights) {
        int start = 0;
        int end = (int) 1e9;
        int ans = end;

        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (possible(heights, mid)) {
                end = mid - 1;
                ans = mid;
            } else {
                start = mid + 1;
            }
        }
        return ans;
    }
}

File: AlgorithmsPackPythonSolutions/Path With Minimum Effort.py
class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        dr = [-1, 0, 1, 0]
        dc = [0, 1, 0, -1]

        def isValid(r, c, heights):
            if r < 0 or r >= len(heights):
                return False
            if c < 0 or c >= len(heights[0]):
                return False
            return True

        def dfs(r, c, heights, visited, diff):
            if r == len(heights) - 1 and c == len(heights[0]) - 1:
                return True

            visited[r][c] = True

            for d in range(4):
                nr = dr[d] + r
                nc = dc[d] + c
                if not isValid(nr, nc, heights) or visited[nr][nc] or abs(heights[nr][nc] - heights[r][c]) > diff:
                    continue

                if dfs(nr, nc, heights, visited, diff):
                    return True

            return False

        def possible(heights, diff):
            rows = len(heights)
            cols = len(heights[0])
            visited = [[False] * cols for _ in range(rows)]

            return dfs(0, 0, heights, visited, diff)

        start = 0
        end = int(1e9)
        ans = end

        while start <= end:
            mid = start + (end - start) // 2
            if possible(heights, mid):
                end = mid - 1
                ans = mid
            else:
                start = mid + 1

        return ans
